interp/src/stk_env.rs

```diff
@@ -0,0 +1,804 @@
+    /// # Examples
+    /// ## Pictoral Example
+    /// ```text
+    /// Say our Smoosher A looked as follows:
+    /// (lvl 3) [(a, 1), (b, 2)] -> [(a, 3)] -> [(c, 4)] -> [(d, 15)] (lvl 0)
+    /// A.diff(3) gives the following HashMap:
+    /// [(a, 3), (c, 4), (d, 15)]
+    /// ```
+    /// ## Code Example
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// a.set("hello", 0);
+    /// let diff1 = a.diff(1); //the diff between the topmost scope and all below
+    /// assert_eq!(diff1.len(), 2);
+    /// assert_eq!(**diff1.get(&"hello").unwrap(), 0);
+    /// assert_eq!(**diff1.get(&"bye").unwrap(), 0);
+    /// ```
+    pub fn diff(&self, levels: u64) -> HashMap<&K, &V> {
+        if levels == 0 {
+            panic!("cannot compute diff(0)");
+        }
+        //iterate from top (0) to (levels - 1) and add all bindings
+        //continue iterating from (levels - 1) to bottom, check if binding
+        //is in tr HM. If so, remove it.
+        let mut tr = HashMap::new();
+        //first add from head
+        for (k, v) in HashMap::iter(&self.head) {
+            tr.insert(k, v);
+        }
+        //now worry about tail. while 1 <= ind <= levels - 1 insert (preserve scope),
+        //and while ind >= levels, get, check, remove.
+        let mut ind = 1;
+        for nd in List::iter(&self.tail) {
+            for (k, v) in HashMap::iter(nd) {
+                if ind <= (levels - 1) {
+                    //add, but only if the binding isn't yet in the HM (preserve scope)
+                    if None == tr.get(k) {
+                        tr.insert(k, v);
+                    }
+                } else {
+                    //do the check and remove
+                    if let Some(&v_prime) = tr.get(k) {
+                        if v_prime == v {
+                            tr.remove(k); //bc it's not a new binding
+                        }
+                    }
+                }
+                ind += 1;
+            }
+        }
+        tr
+    }
+
+    /// ```text
+    /// Returns a HM of all (&K, &V) (bindings of references) found in [self].
+    /// A use case would be when you want a HM representing a snapshot of the
+    /// current state of the environment, which is easily iterable.
+    /// ```
+    pub fn to_hm(&self) -> HashMap<&K, &V> {
+        //just add all bindings
+        let mut tr = HashMap::new();
+        //first add from head
+        for (k, v) in HashMap::iter(&self.head) {
+            tr.insert(k, v);
+        }
+        //then from tail
+        for nd in List::iter(&self.tail) {
+            for (k, v) in HashMap::iter(nd) {
+                //add, but only if the binding isn't yet in the HM (preserve scope)
+                if None == tr.get(k) {
+                    tr.insert(k, v);
+                }
+            }
+        }
+        tr
+    }
```

Comment:
Better to implement as Into, unless there's a reason not to use the trait here

Reply:
Into consumes the input, we don't want it to consume

Reply:
implement it for `&Smoosher` not for `Smoosher`